File: LeetCode/MinDepthOfBinaryTree.java
import java.util.*;

/**
 * Created by yi on 4/24/15.
 */
public class MinDepthOfBinaryTree {
    public int minDepth(TreeNode root) {
        Queue<TreeNode> que = new LinkedList<TreeNode>();
        if (root == null) return 0;
        TreeNode rightMost = root;
        int level = 0;
        que.add(root);
        while (!que.isEmpty()) {
            TreeNode node = que.poll();
            if (node.left != null)
                que.add(node.left);
            if (node.right != null)
                que.add(node.right);
            if (node.right == null && node.left == null)
                break;
            if (node == rightMost) {
                rightMost = node.right;
                level++;
            }
        }
        return level;
    }

    private class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            this.val = x;
        }
    }
}

File: LeetCode/Python/MinDepthOfBinaryTree.py
from collections import deque

__author__ = 'yi'


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def minDepth(self, root):
        store = deque([root])
        right_most = root
        level = 1
        while store is not None:
            copy = store.popleft()
            if copy.right is None and copy.left is None:
                break
            if copy.left is not None:
                store.append(copy.left)
            if copy.right is not None:
                store.append(copy.right)
            if copy == right_most:
                level += 1
                right_most = copy.right
        return level


if __name__ == '__main__':
    root = TreeNode(input())
    print(Solution().minDepth(root))